deepchem/feat/base_classes.py

```diff
@@ -6,16 +6,60 @@
 import numpy as np
 import multiprocessing
 
-__author__ = "Steven Kearnes"
-__copyright__ = "Copyright 2014, Stanford University"
-__license__ = "BSD 3-clause"
+logger = logging.getLogger(__name__)
 
 
 def _featurize_complex(featurizer, mol_pdb_file, protein_pdb_file, log_message):
   logging.info(log_message)
   return featurizer._featurize_complex(mol_pdb_file, protein_pdb_file)
 
 
+class Featurizer(object):
+  """Abstract class for calculating a set of features for a datapoint.
+
+  This class is abstract and cannot be invoked directly. You'll
+  likely only interact with this class if you're a developer. In
+  that case, you might want to make a child class which
+  implements the `_featurize` method for calculating features for
+  a single datapoints if you'd like to make a featurizer for a
+  new datatype.
+  """
+
+  def featurize(self, datapoints, log_every_n=1000):
+    """Calculate features for datapoints.
+
+    Parameters
+    ----------
+    datapoints: object 
+       Any blob of data you like. Subclasss should instantiate this. 
```

Comment:
The implementation doesn't match the description.  It passes `datapoints` to `list()`, which requires that it be iterable.  Then it treats that list as a sequence of separate datapoints and featurizes each one separately.  So consider the example you gave of calling `featurizer("CC")`.  It would consider that to be a sequence of two datapoints each consisting of the single character "C", and featurize each one independently.

What is the intended behavior: that you pass in a single datapoint, or a sequence of datapoints?

Reply:
Ah good point. I think the idea is that if the input is a sequence, it just uses it directly. If not, then it wraps the input in a list and processes it. So we can do

```
x = feat("CC")
xs = feat(["C", "CC", "CO"])
````

as both valid usage

Reply:
The difficulty is that so many of the objects we deal with, like strings and numpy arrays, are already sequences.  This creates a lot of opportunities for bugs.

Reply:
Ah you're right. Let me just make this require a sequence of objects to featurize to simplify the possibility for bugs.